updated contract calls page (#701)
multiversx/mx-docs: docs/developers/developer-reference/sc-contract-calls.md

We call this contract call helper a __proxy__. All it does is that it provides a typed interface to any Rust program, it takes the typed arguments, and it serializes them according to the [MultiversX serialization format](/developers/data/serialization-overview).

Let's take this very simple example:

```rust
adder_proxy.add(3u32)
```

Here, we have a proxy to the adder contract. The `add` method doesn't call the smart contract directly. Instead, it produces a contract call object that contains the data field `add@03`, which is something that the blockchain can make sense of. We will see later how this contract call can end up actually being executed. But until then, let's see how we can get hold of one of these proxies.

This guide provides some examples on how to call a contract from another contract. More examples can be found in [the contract composability feature tests](https://github.com/multiversx/mx-sdk-rs/tree/master/contracts/feature-tests/composability).

There are three ways of doing these calls:

> - importing the callee contract's source code and using the auto-generated proxy (recommended)

Author: A new way of creating proxies is coming soon, will be the recommended way. Writing them by hand is not necessarily the way.

:::

---

[comment]: # (mx-context-auto)

## Contract calls: base

Smart contract calls at the blockchain level have no notion of arity, or data types. That is, the blockchain itself does not validate the number of arguments (or results), and each of these only appear as raw binary data fields.

It is the contract that keeps track of the number of arguments, and deserializes them. If a transaction has the wrong number of arguments, it is only the contract itself that will be able to complain. If the types are off, it is only during deserialization that the contract will know.

The description of a smart contract's inputs is known as the [ABI](/developers/data/abi), and lives off-chain. In short, the ABI is a collection of endpoint names, with argument names and type descriptions. To be able to effectively call a smart contract, it is useful to know its ABI.

> We call this contract call helper a __proxy__. All it does is that it provides a typed interface to any Rust program, it takes the typed arguments and it serializes them according to the [MultiversX serialization format](/developers/data/serialization-overview).

Reviewer: For extra clarity:

```suggestion
Knowing the ABI of a contract enables us to call it cleanly. The ABI is used to create a __proxy__ for the contract, which is a helper component that stands in for the contract in our code. All it does is that it provides a typed interface to any Rust program, it takes the typed arguments and it serializes them according to the [MultiversX serialization format](/developers/data/serialization-overview).
```

Author: Not completely correct. The proxy right now is not created based on the ABI.

Author: I rephrased it a little differently.

- git branch, tag or commit.

Relative paths are the most common for contracts in the same workspace.
:::

If the contract has modules with functionality that you may want to call, you will also need to import those.

If the modules are in different crates than the target contract (and if the target contract doesn't somehow re-export them), you'll also have to add the module to the dependencies, the same way you added the target contract.

These proxies are traits, just like the contracts themselves. The implementation is produced automatically, but nonetheless, this means that in order to call them, the proxy trait must be in scope. This is why you will see such imports everywhere these proxies are called:

```rust
use module_namespace::ProxyTrait as _;
```

If you use the [rust-analyser VSCode extension](https://marketplace.visualstudio.com/items?itemName=rust-lang.rust-analyzer), it might complain that it can't find this, but if you actually build the contract, the compiler can find it just fine.

Once you've imported the contract and any external modules it might use, you have to declare a proxy creator function in the contract:
```rust
#[proxy]

> fn callee_contract_proxy(&self, sc_address: ManagedAddress) -> contract_namespace::Proxy<Self::Api>;

Reviewer 2: Below you used the proxy with `callee_sc_address`. Please rename here also

```

This function doesn't do much, it just tries to sort out the proxy trait imports, and neatly initializes the proxy for you.

This function creates an object that contains all the endpoints of the callee contract, and it handles the serialization automatically.

Let's say you have the following endpoint in the contract you wish to call:

```rust
#[endpoint(myEndpoint)]

> fn callee_endpoint(&self, arg: BigUint) -> BigUint {

Reviewer 2: Also, modify the name of the endpoint

}
```

To call this endpoint, you would write something like:

```rust
self.contract_proxy(callee_sc_address)
	.my_endpoint(my_biguint_arg)
	.async_call()
	.call_and_exit();
```

We'll talk about `async_call` and `call_and_exit` later on.

:::caution
Importing a smart contract crate only works if both contracts use the exact framework version. Otherwise, the compiler will (rightfully) complain that the interfaces do not perfectly match.

In case the target contract is not under our control, it is often wiser to just write the proxy of interest by hand.
:::

[comment]: # (mx-context-auto)

> ### Manually specified proxies

Author: Not sure if composed is the better word here. It suggests smaller bits were put together, but these proxies are written from scratch.

Below we have an example of such a proxy:

```rust
mod callee_proxy {
    multiversx_sc::imports!();

    #[multiversx_sc::proxy]
    pub trait CalleeContract {
        #[payable("*")]
		#[endpoint(myEndpoint)]
		fn my_payable_endpoint(&self, arg: BigUint) -> BigUint;
    }
}
```

The syntax is almost the same as for contracts, except the endpoints have no implementation.

:::caution
Just like with smart contracts and modules, proxy declarations need to reside in their own module. This can either be a separate file, or an explicit declaration, like `mod callee_proxy {}` above.

This is because there is quite a lot of code generated in the background that would interfere otherwise.
:::

Manually declared proxies are no different from the auto-generated ones, so calling them looks the same.

```rust
#[proxy]
fn contract_proxy(&self, sc_address: ManagedAddress) -> callee_proxy::Proxy<Self::Api>;
```

[comment]: # (mx-context-auto)

### No proxy

The point of the proxies is to help us build contract calls in a type-safe manner. But this is by no means compulsory. Sometimes we specifically want to build contract calls by hand and serialize the arguments ourselves.

> The point is to create a `ContractCallNoPayment` object. We'll discuss how to add the payments later.

Reviewer 2: In order to do that, we want to create a [...]

The `ContractCallNoPayment` has two type arguments: the API and the expected return type. If we are in a contract, the API will always be the same as for the entire contract. To avoid having to explicitly specify it, we can use the following syntax:

```rust
let mut contract_call = self.send()
	.contract_call::<ResultType>(to, endpoint_name);
contract_call.push_raw_argument(arg1_encoded);
contract_call.push_raw_argument(arg2_encoded);
```

> If we are not in a smart contract we might do the same thing using the same, equivalent syntax:

Reviewer 2: For me it's a bit ambiguous. Can you rephrase?

```rust
let mut contract_call = ContractCallNoPayment::<StaticApi, ResultType>::new(to, endpoint_name);
contract_call.push_raw_argument(arg1_encoded);
contract_call.push_raw_argument(arg2_encoded);
```

<!-- TODO: we need slightly better syntax for the arguments. Was overlooked because we never used them like this -->

[comment]: # (mx-context-auto)

### Diagram

Up to here we have created a contract call without payment, so an object of type `ContractCallNoPayment`, in the following ways:

```mermaid
graph LR
    gen-proxy[Generated Proxy] --> ccnp[ContractCallNoPayment]
    man-proxy[Manual Proxy] --> ccnp
    ccnp-new["ContractCallNoPayment::new(to, function)"] --> ccnp

> ```

Reviewer 2: Can you center the image?

```mermaid
graph LR
    ccnp[ContractCallNoPayment]
	ccnp -->|".with_egld_transfer"| cc-egld[ContractCallWithEgld]
	ccnp -->|".with_esdt_transfer"| cc-multi[ContractCallWithMultiEsdt]
	cc-multi -->|".with_esdt_transfer"| cc-multi
	ccnp -->|".with_multi_token_transfer"| cc-multi
	ccnp -->|".with_any_payment"| cc-any[ContractCallWithAnyPayment]
	ccnp -->|".with_egld_or_single_esdt_transfer"| cc-egld-single[ContractCallWithEgldOrSingleEsdt]
```

---

[comment]: # (mx-context-auto)

## Contract calls: gas

Specifying gas is fairly straightforward. All contract call objects have a `with_gas_limit` method, so gas can be specified at any point.

Not all contract calls require explicit specification of the gas limit, leaving it out is sometimes fine. Notably:
- async calls will halt execution and consume all the remaining gas, so specifying the gas limit is not necessary for them;
- synchronous calls will by default simply use all the available gas as the upper limit, since unspent gas is returned to the caller anyway.

On the other hand,promises and transfer-execute calls do require gas to be explicitly specified.

> 

Reviewer 2: Add snippets of code with `with_gas_limit` as an example.

---

[comment]: # (mx-context-auto)

## Contract calls: execution

There are several ways in which contract calls are launched from another contract. Currently they are:
- asynchronous calls:
	- single asynchronous calls:

> 	- promises (multi ple asynchronous calls),

Reviewer 2: `multi ple` -> multiple

### Callbacks

But before sending the asynchronous call, we most likely will want to also configure a callback for it. This is the only way that our contract will be able to react to the outcome of its execution.

> Let's imagine that `my_endpoint` returns a `BigUint`, and we want to do something if the result is even, and something else if the result is odd. We also want to do some cleanup in case of error. Our callback function would look something like this:

Reviewer 2: `my_endpoint` -> `caller_endpoint`

}
```

The `#[call_result]` argument interprets the output of the called endpoint and must almost always be of type `ManagedAsyncCallResult`. This type decodes the error status from the VM, more about it [here](/developers/data/multi-values#standard-multi-values). Its type argument must match the return type of the called endpoint.

To assign this callback to the aforementioned async call, we hook it after `async_call`, but before `call_and_exit`:

```rust
#[endpoint]
fn caller_endpoint(&self) {
	// previous code here

	self.callee_contract_proxy(callee_sc_address)
		.callee_endpoint(my_biguint_arg)
		.async_call()

> 		.with_callback(self.callbacks().callee_endpoint_callback())

Reviewer: `callee_endpoint_callback` is called here, should it rather be a function pointer? Without the parantheses?

Author: I wish we had the power to redefine function pointers. We don't.

The next best thing: fooling the developer. This is no the actual callback, it is the callback proxy, that function actually produces a `CallbackCall`.

I won't be writing about it today. Maybe another day.

Marvel in the mischief!

:::

Callbacks can also receive payments, both EGLD and ESDT. They are always payable, there is never any need to annotate them with ``#[payable]`. They will receive payments if the called contract sends back tokens to the caller. In this case, they can query the received payments, just like a regular payable endpoint would.

```rust
#[callback]
fn my_endpoint_callback(&self, #[call_result] result: ManagedAsyncCallResult<BigUint>) {
	let payment = self.call_value().any_payment();

	// ...
}
```

:::note Note on implementation
Even though, in theory, smart contract can only have ONE callback function, the Rust framework handles this for you by saving an ID for the callback function in storage when you fire the async call, and it knows how to retrieve the ID and call the correct function once the call returns.
:::

[comment]: # (mx-context-auto)

### The callback closure

Assume there is some additional context that we want to pass from our contract directly to the callback. We cannot rely on the called contract to do the job for us, we want to be in full control of this context.

This context forms the contents of our callback closure.

> More specifically, all callback arguments other than the `#[call_result]` will be passed to it before launching the call, they will be saved by the framework somewhere in the contract storage, then given to the callback and deleted.

Reviewer 2: remove `somewhere`

```

Notice how the callback now has an argument:

```rust
self.callbacks().my_endpoint_callback(caller)
```

You can then use `original_caller` in the callback like any other function argument.

[comment]: # (mx-context-auto)

### Promises

> Promises (or multi-async calls) are a new feature that will be introduced in mainnet release 1.6. They are very similar to the old asynchronous calls, with the difference that launching them does not terminate current execution, and there can be several launched from the same transaction.

Reviewer 2: ~are a feature that are available from the mainnet release 1.6.~

Reviewer 2: Also, a suggestion: programmers (technical people) like bullets. You can create 2 bullets here.

:::caution

> Because this feature is currently not available on mainnet, contracts need to enable the "promises" feature flag in `Cargo.toml` to use the functionality:

Reviewer 3: Don't know what would be the best solution here to avoid the need to modify the docs after the release of 1.6

Author: We will modify them.

}
```

The differences are:
- Calling `async_call_promise` instead of `async_call`.
- Calling `register_promise` instead of `call_and_exit`. This one does not terminate execution.
- Annotation `#[promises_callback]` instead of `#[callback]`.
- We need to specify the gas for the call, because the execution of our transaction will continue and it needs to know how much gas it can keep.
- We need to specify the amount of gas for the callback. This is the exact amount of gas reserved for the callback, irrespective of how much the target contract consumes.

[comment]: # (mx-context-auto)

### Transfer-execute

> Transfer-execute calls are similar to asynchronous calls, but they can have no callback, and thus the caller cannot react in any way to what happens with the callee.

Reviewer 3: ```suggestion
Transfer-execute calls are similar to asynchronous calls, but they can have no callback, thus the caller cannot react in any way to what happens with the callee.
```

> Just like promises, there can be multiple such calls launched from a transaction, but unlike promises, they are alreay available on mainnet.

Reviewer 3: ```suggestion
Just like promises, there can be multiple such calls launched from a transaction, but unlike promises, these are already available on mainnet.
```
Same question here regarding the need to modify afterwards.

> Transfer-execute calls do not need any further configuration (other than explicit an explicit gas limit), and therefore there is no specific object type associated with them. They can be launched immediately:

Reviewer 3: ```suggestion
Transfer-execute calls do not need any further configuration (other than an explicit gas limit), therefore there is no specific object type associated with them. They can be launched immediately:
```

```rust
self.contract_proxy(callee_sc_address)
	.my_endpoint(my_biguint_arg)
	.with_gas_limit(gas_limit)
	.transfer_execute();
```

[comment]: # (mx-context-auto)

### Synchronous calls

> Synchronous calls are executed inline: this means execution is interrupted while they are executed, and resumed afterwards. We also get the result of the execution right away and we can use it immediately in the transaction.

Reviewer 3: ```suggestion
Synchronous calls are executed inline: this means execution is interrupted while they are executed and resumed afterwards. We also get the result of the execution right away and we can use it immediately in the transaction.
```

Reviewer 2: I would add caller and callee to the explaination.
`this means the caller execution is interrupted while the callee function is executed`

> The catch is that synchronous calls can only be sent to contracts in the __same shard__ as the caller. They will fail otherwise.

Reviewer 3: ```suggestion
:::caution
Synchronous calls can only be sent to contracts in the __same shard__ as the caller. They will fail otherwise.
:::
```

    cc ---->|".execute_on_dest_context()
              .execute_on_dest_context_readonly()
              .execute_on_same_context()"| exec-dest["⚙️ Synchronous call"]
    exec-dest --> result[Requested Result]
```

---

[comment]: # (mx-context-auto)

## Contract deploy: base

A close relative of the contract call is the contract deploy call. It models the deployment or the upgrade of a smart contract.

It shares a lot in common with the contract calls, with these notable differences:
- The endpoint name is always `init`.
- No ESDT transfers are allowed in `init`.
- They get executed slightly differently.

```mermaid
flowchart TB
    gen-proxy[Generated Proxy] --> cc
    man-proxy[Manual Proxy] --> cc
    ccnp-new["ContractDeploy::new"] --> cc
    cc[ContractDeploy]
    cc ---->|".deploy_contract()
              .deploy_from_source()"| exec-deploy["⚙️ Deploy contract"]
    exec-deploy --> result[Requested Result]
    cc ---->|".deploy_contract()
              .deploy_from_source()"| exec-upg["⚙️ Upgrade contract"]
```

The object encoding these calls is called `ContractDeploy`. Unlike the contract calls, there is a single such object.

> Creating this object is done in a similar fashion: either via proxies, or by hand. Constructors in proxies naturally produce `ContractDeploy` objects:

Reviewer 3: ```suggestion
Creating this object is done in a similar fashion: either via proxies, or manually. Constructors in proxies naturally produce `ContractDeploy` objects:
```

```rust
mod callee_proxy {
    multiversx_sc::imports!();

    #[multiversx_sc::proxy]
    pub trait CalleeContract {
		#[init]
		fn init(&self, arg: BigUint) -> BigUint;
    }
}
```

```rust
self.contract_proxy()
	.init(my_biguint_arg)
```

---

[comment]: # (mx-context-auto)

## Contract deploy: configuration

Just like with regular contract calls, we can specify the gas limit and perform EGLD transfers as part of the deploy as follows:

```rust
self.contract_proxy(callee_sc_address)
	.init(my_biguint_arg)

```

```rust
self.contract_proxy()
	.init(my_biguint_arg)

> 	.with_egld_transfer(egld_amount)

Reviewer 2: is `with_egld_transfer` available on deploy? I think this is an upgrade, right?

Author: It is!

```rust
let (new_address, result)  = self.contract_proxy()
	.init(my_biguint_arg)
	.deploy_contract::<ResultType>(code, code_metadata);
```

:::important
Contract deploys always happen in the same shard as the deployer. They are therefore always [synchronous calls](#synchronous-calls) and we get the result right away. Just like for `execute_on_dest_context` we need to either write a result with an explicit result type, or give the result type as type argument.

Contract upgrades, on the other hand, can be sent to a different shard, and are therefore in essence asynchronous calls.
:::

The methods for executing contract deploys are as follows:
- `.deploy_contract(code, code_metadata)` - deploys a new contract with the code given by the contract.
- `.deploy_from_source(source_address, code_metadata)` - deploys a new contract with the same code as the code of the contract at `source_address`. The advantage is that the contract doesn't need to handle the new contract code, which could be quite a large data blob. This saves gas. It requires that we have the code already deployed somewhere else.

[comment]: # (mx-context-auto)

> ### Upgrade

Reviewer 3: In the next mainnet release we'll have deffered inits for upgrades. We'll have to care for this as well very soon. Maybe I more places.

Author: I know ....
But first we need the implementation, then the docs ....
